### ISSUE TITLE
Trouble to understand position embedding.

### ISSUE THREAD

**Issue by guotong1988** [google-research/bert#58]
![image](https://user-images.githubusercontent.com/4702353/48039361-6a13a980-e1af-11e8-9f61-38b8a986080b.png)

`position_embeddings` is only a matrix which is random init?

Which code part means the position info of word in sentence?

Thank you!!

**Comment by jacobdevlin-google** [comment]
It was randomly initialized when it was created but it was trained during pre-training with the rest of the network. You don't need to store the position info. `full_position_embeddings` is a tensor of shape `[max_position_embeddings, width]`. So `full_position_embeddings[i:i+1,]` is the position embedding of position `i`. So you can just add it to the input matrix and each item will be applied at the correct position (which is what's done right after that code).

**Comment by guotong1988** [comment]
Thank you very much.
It hard to understand why it works.

**Comment by jacobdevlin-google** [comment]
Well if you don't have a positional embedding matrix then the Transformer has no way of knowing the relative position of each word. It would be exactly like randomly shuffling the input sentence. So the positional embeddings let the model learn the actual sequential ordering of the input sentence (which something like an LSTM gets for free).

**Comment by guotong1988** [comment]
Thank you very much, again.

**Comment by guotong1988** [comment]
![image](https://user-images.githubusercontent.com/4702353/48047558-57ab6700-e1d3-11e8-97c3-662cd109d8d1.png)

I still don't know why this op can let it work like LSTM.

**Comment by jacobdevlin-google** [comment]
This is a general technique used in the Transformer, which has been successfully in hundreds of papers, so it's outside the scope of this page. Please see a general guide to the Transformer such as [The Illustrated Transformer](http://jalammar.github.io/illustrated-transformer/) if you want to know more about it. They have a section on the positional embeddings with a heatmap.

**Comment by mosicr** [comment]
It seems like position embedding is not properly implemented in Google Bert Python. It is reinitialized on each pass; there are no sine / cosine positional updates as per Section 3.5 of Attention is All You Need . 
https://medium.com/@ranko.mosic/googles-bert-nlp-5b2bb1236d78

**Comment by kapilkd13** [comment]
Hi @jacobdevlin-google I have small doubt, b default the `max_position_embeddings` arg is set to 512 in bert_config file of the downloaded model (Cased-12 Layer model open-sourced by google).
Now I want to send a sequence larger than 512 tokens. Can I do this in any way?
What if I change the `max_position_embeddings` value in `config` file to something like 2048. 
Or is this variable dependent on the trained model?

**Comment by mealsd** [comment]
It is a little confused for me in two aspects.
1. the Transformer does not shuffle the input sequence, so the sequence information could be retained. the multi-head attent is just related to the length of embedding rather than the sequence.
2. when the sequence is too long, whether this position embedding works or not.  If I set the max_sequence too large will that be a potential problem?

**Comment by hsm207** [comment]
@mealsd the largest max_sequence you can set is 512 because this was the value used during pretraining. So, there's no problem as long as your max_sequence length is less than 512.

**Comment by mealsd** [comment]
> @mealsd the largest max_sequence you can set is 512 because this was the value used during pretraining. So, there's no problem as long as your max_sequence length is less than 512.

Actually, I'm doing something on a Transformer structure, and maybe  need 9000 length position embedding.  I'm confusing whether this make sense or not.

**Comment by hsm207** [comment]
If your input sequence is a really long text e.g. an essay vs a sentence, then 9k length position embedding makes sense. Unfortunately, BERT is only able to handle sequence of up to length 512. As an alternative, you may consider chunking your input sequence into smaller units e.g. paragraphs.

**Comment by bnicholl** [comment]
Isn't the positional embedding supposed to be a function of sine / cosine. @jacobdevlin-google say's this is randomly initialized. Does BERT initialize the positional embedding differently?

**Comment by hsm207** [comment]
@bnicholl in BERT, the positional embedding is a learnable feature. As far as I know, the sine/cosine thing was introduced in the[ attention is all you need](https://arxiv.org/pdf/1706.03762.pdf) paper and they found that it produces almost the same results as making it a learnable feature:

![image](https://user-images.githubusercontent.com/2398765/72021353-c50bc980-32a8-11ea-8caf-84bd06aef60d.png)

**Comment by bnicholl** [comment]
Thanks for the response. Last question. Are the token embeddings the integer values I obtain when I call:
`tokenizer = BertTokenizer.from_pretrained('bert-base-uncased', do_lower_case=True) `
`print('Ids: ', tokenizer.encode( string_values_from_sentence ))`

`OUTPUT:  Ids:  [101, 2183, 2000, 15132, 2023, 5353, 102]`

If so, I'm assuming these tokens are randomly initialized, and are not learned(since there of type int)?

**Comment by hsm207** [comment]
@bnicholl no, `[101, 2183, 2000, 15132, 2023, 5353, 102]` are not the token embeddings.

They are the index to the token embeddings.

There's a huge matrix with dimension vocab size x embedding dimension  that records the embedding for each token. So, to get the embedding for the word "going", you will go to row 101 of this matrix (assuming "going" gets tokenized to "going" instead of "go" and "ing".)

Hope this clarifies.

**Comment by bnicholl** [comment]
Thanks for clarifying. So in the example above, if the word going gets broken to ##go ##ing, how does that word embedding get learned. It is my understanding that if the word is not in the corpus, that is when the word gets split into multiple words. So, if BERT comes across a word that it is not familiar with during the fine tuning phase, does it just randomly initialize that word with a word embedding, then minimize the error during fine tuning?

**Comment by hsm207** [comment]
Words that are unknown to BERT will get replace with a special token. I think its [UNK] (not entirely sure, since it's been awhile since I work with BERT). [UNK] tokens will have their own embedding that is learned too during the finetuning phase.

**Comment by linhlt-it-ee** [comment]
> @bnicholl in BERT, the positional embedding is a learnable feature. As far as I know, the sine/cosine thing was introduced in the[ attention is all you need](https://arxiv.org/pdf/1706.03762.pdf) paper and they found that it produces almost the same results as making it a learnable feature:
> 
> ![image](https://user-images.githubusercontent.com/2398765/72021353-c50bc980-32a8-11ea-8caf-84bd06aef60d.png)

Tks for clarifying. However I think they choose the learned position embedding because it would dramatically change corresponding with the change of context (words around) when fine tuning.

**Comment by BaoshengHeTR** [comment]
> > @bnicholl in BERT, the positional embedding is a learnable feature. As far as I know, the sine/cosine thing was introduced in the[ attention is all you need](https://arxiv.org/pdf/1706.03762.pdf) paper and they found that it produces almost the same results as making it a learnable feature:
> > ![image](https://user-images.githubusercontent.com/2398765/72021353-c50bc980-32a8-11ea-8caf-84bd06aef60d.png)
> 
> Tks for clarifying. However I think they choose the learned position embedding because it would dramatically change corresponding with the change of context (words around) when fine tuning.

Well. That looks reasonable at the first glance.
However, sinusoidal functions have many good features, e.g., the different varying period at different dimensions. I don not think using a randomly initialized embedding can bring the same benefits.

**Comment by lizzy2689** [comment]
With the huge dataset, it could be. I thought like you at first. But BERT code does not lie. If you have more discoveries, please share with me. Tks

**Comment by linhlt-it-ee** [comment]
> > > @bnicholl in BERT, the positional embedding is a learnable feature. As far as I know, the sine/cosine thing was introduced in the[ attention is all you need](https://arxiv.org/pdf/1706.03762.pdf) paper and they found that it produces almost the same results as making it a learnable feature:
> > > ![image](https://user-images.githubusercontent.com/2398765/72021353-c50bc980-32a8-11ea-8caf-84bd06aef60d.png)
> > 
> > 
> > Tks for clarifying. However I think they choose the learned position embedding because it would dramatically change corresponding with the change of context (words around) when fine tuning.
> 
> Well. That looks reasonable at the first glance.
> However, sinusoidal functions have many good features, e.g., the different varying period at different dimensions. I don not think using a randomly initialized embedding can bring the same benefits.
With the huge dataset, it could be. I think like you at first. However, in BERT CODE I trust. :). Please share with me if you discover something else. Tks

**Comment by BaoshengHeTR** [comment]
> > > > @bnicholl in BERT, the positional embedding is a learnable feature. As far as I know, the sine/cosine thing was introduced in the[ attention is all you need](https://arxiv.org/pdf/1706.03762.pdf) paper and they found that it produces almost the same results as making it a learnable feature:
> > > > ![image](https://user-images.githubusercontent.com/2398765/72021353-c50bc980-32a8-11ea-8caf-84bd06aef60d.png)
> > > 
> > > 
> > > Tks for clarifying. However I think they choose the learned position embedding because it would dramatically change corresponding with the change of context (words around) when fine tuning.
> > 
> > 
> > Well. That looks reasonable at the first glance.
> > However, sinusoidal functions have many good features, e.g., the different varying period at different dimensions. I don not think using a randomly initialized embedding can bring the same benefits.
> 
> With the huge dataset, it could be. I think like you at first. However, in BERT CODE I trust. :). Please share with me if you discover something else. Tks

with very large dataset, you can achieve such good staff? Like local periodcial and linear consistency? Hope people from google can give an explanation.

**Comment by dgg32** [comment]
So please correct me whether I understand BERT embedding correctly please:

position embedding is a matrix with a shape of 512 x 768. 

512 is the length that BERT can take, defined in the config file.

768 is the word embedding vector length.

The values inside this matrix were learned during the pretraining.

If I have two batches of inputs, the first words in both batches will get the same first x768 values out of the 512 (the first row of 512). The second words in both batches will get the same second x768 values of the 512 (the second row of 512), and so on.

For each word, BERT will do simple 768 element by element additions for "word embedding + segment embedding + position embedding". The result is the embedding layer.

Is my understanding correct please? Thank you!

**Comment by vigneshgig** [comment]
> This is a general technique used in the Transformer, which has been successfully in hundreds of papers, so it's outside the scope of this page. Please see a general guide to the Transformer such as [The Illustrated Transformer](http://jalammar.github.io/illustrated-transformer/) if you want to know more about it. They have a section on the positional embeddings with a heatmap.

Hi, I have one doubt regarding position embedding. I have seen the given link, but I am unable to find positional embedding. i can find only positional encoding. I clearly understand the positional encoding purpose. 
But how exactly do we know positional embedding is learning position because for token embedding layer at least we giving an input. but for positional embedding, we just do concat with the given token embedding. I am still unable to find any proper explanation about positional embedding. please help me out

**Comment by kechan** [comment]
Anyone know here the reason Bert had preferred to learn the positional embeddings instead of using a fixed sin/cos ones as in “Attn is all you need”? An obvious guess would be better performance in the overall downstream task (which I assumed will also fine-tuned those embeddings). The Bert paper I don’t think had said too much about this. In fact, I was only surprised when i cared to go to read the Appendix about expensive training for long sequence. 

I wonder if the learnt embeddings resulting from those downstream task (or pre-trained), will still have the “nice” properties for fixed norm, relative position agnostic properties, etc. etc. I ain’t surprised if they violate some of them, again showing manual feature engineering is trumped by just learning them from massive data.
